### PR TITLE
schedule: fix Scatter's ordinary filter causing data race

### LINKED ISSUE #5317: pd panic after run tikv down test
### What did you do?
OLTP_STABILITY_REL_001
[2022/07/18 09:21:39.873 +08:00] [INFO] [chaos.go:358] ["fault will last for"] [duration=20m0s]
[2022/07/18 09:21:39.877 +08:00] [INFO] [chaos.go:93] ["Run chaos"] [name=failure] [selectors="[endless-oltp-tps-990456-1-718/tc-tikv-1]"] [SelectorsRetainPolicy(selectors)="[endless-oltp-tps-990456-1-718/tc-tikv-1]"] [targetSelectors="[nil]"] [TargetSelectorsRetainPolicy(targetSelectors)="[nil]"] [experimentSpec="FailureExperimentSpec{Duration: \"\", Scheduler: <nil>}"]

### What did you expect to see?
pd panic
```
2022-07-18 09:27:05	log="nslookup domain [tc-pd-0.tc](http://tc-pd-0.tc/)-pd-peer.endless-oltp-tps-990456-1-718.svc.svc success\n"
2022-07-18 09:27:05	log="\n"
2022-07-18 09:27:05	log="Address: [10.233.79.21](http://10.233.79.21/)\n"
2022-07-18 09:27:05	log="Name:\[ttc-pd-0.tc-pd-peer.endless-oltp-tps-990456-1-718.svc.cluster.local](http://ttc-pd-0.tc-pd-peer.endless-oltp-tps-990456-1-718.svc.cluster.local/)\n"
2022-07-18 09:27:05	log="\n"
2022-07-18 09:27:05	log="Address:\t10.96.0.10#53\n"
2022-07-18 09:27:05	log="Server:\t\t10.96.0.10\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[google.golang.org/grpc@v1.26.0/server.go:720](http://google.golang.org/grpc@v1.26.0/server.go:720) +0xea\n"
2022-07-18 09:27:03	log="created by [google.golang.org/grpc.(*Server).serveStreams.func1\n](http://google.golang.org/grpc.(*Server).serveStreams.func1/n)"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[google.golang.org/grpc@v1.26.0/server.go:722](http://google.golang.org/grpc@v1.26.0/server.go:722) +0x98\n"
2022-07-18 09:27:03	log="[google.golang.org/grpc.(*Server).serveStreams.func1.1()\n](http://google.golang.org/grpc.(*Server).serveStreams.func1.1()/n)"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[google.golang.org/grpc@v1.26.0/server.go:1313](http://google.golang.org/grpc@v1.26.0/server.go:1313) +0xa05\n"
2022-07-18 09:27:03	log="[google.golang.org/grpc.(*Server).handleStream](http://google.golang.org/grpc.(*Server).handleStream)(0xc000fd0f00, {0x33efb10, 0xc002ddf980}, 0xc002c7f700, 0x0)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[google.golang.org/grpc@v1.26.0/server.go:1024](http://google.golang.org/grpc@v1.26.0/server.go:1024) +0xd1e\n"
2022-07-18 09:27:03	log="[google.golang.org/grpc.(*Server).processUnaryRPC](http://google.golang.org/grpc.(*Server).processUnaryRPC)(0xc000fd0f00, {0x33efb10, 0xc002ddf980}, 0xc002c7f700, 0xc000f6cf00, 0x4382150, 0x0)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/pingcap/kvproto@v0.0.0-20220510035547-0e2f26c0a46a/pkg/pdpb/pdpb.pb.go:7693](http://github.com/pingcap/kvproto@v0.0.0-20220510035547-0e2f26c0a46a/pkg/pdpb/pdpb.pb.go:7693) +0x138\n"
2022-07-18 09:27:03	log="[github.com/pingcap/kvproto/pkg/pdpb._PD_ScatterRegion_Handler](http://github.com/pingcap/kvproto/pkg/pdpb._PD_ScatterRegion_Handler)({0x2721d80?, 0xc000010538}, {0x33e4e10, 0xc003effec0}, 0xc0096257a0, 0xc00146a8d0)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:39](http://github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:39) +0x1a3\n"
2022-07-18 09:27:03	log="[github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1](http://github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1)({0x33e4e10, 0xc003effec0}, {0x26bbf80, 0xc0029449a0}, 0xc006bed8e0, 0xc002e79ed8)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:71](http://go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:71) +0xc3\n"
2022-07-18 09:27:03	log="[go.etcd.io/etcd/etcdserver/api/v3rpc.newLogUnaryInterceptor.func1](http://go.etcd.io/etcd/etcdserver/api/v3rpc.newLogUnaryInterceptor.func1)({0x33e4e10, 0xc003effec0}, {0x26bbf80, 0xc0029449a0}, 0xc006bed8e0, 0xc000240f00)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34](http://github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34) +0x6f\n"
2022-07-18 09:27:03	log="[github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1](http://github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1)({0x33e4e10?, 0xc003effec0?}, {0x26bbf80?, 0xc0029449a0?})\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:64](http://go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:64) +0x1b6\n"
2022-07-18 09:27:03	log="[go.etcd.io/etcd/etcdserver/api/v3rpc.newUnaryInterceptor.func1](http://go.etcd.io/etcd/etcdserver/api/v3rpc.newUnaryInterceptor.func1)({0x33e4e10, 0xc003effec0}, {0x26bbf80?, 0xc0029449a0}, 0x0?, 0xc000240f00)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34](http://github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34) +0x6f\n"
2022-07-18 09:27:03	log="[github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1](http://github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1)({0x33e4e10?, 0xc003effec0?}, {0x26bbf80?, 0xc0029449a0?})\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107](http://github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107) +0x87\n"
2022-07-18 09:27:03	log="[github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1](http://github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1)({0x33e4e10, 0xc003effec0}, {0x26bbf80, 0xc0029449a0}, 0x24cab20?, 0xc000240f00)\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:31](http://github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:31) +0x89\n"
2022-07-18 09:27:03	log="[github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1](http://github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1)({0x33e4e10?, 0xc003effec0?}, {0x26bbf80?, 0xc0029449a0?})\n"
2022-07-18 09:27:03	log="\t/go/pkg/mod/[github.com/pingcap/kvproto@v0.0.0-20220510035547-0e2f26c0a46a/pkg/pdpb/pdpb.pb.go:7691](http://github.com/pingcap/kvproto@v0.0.0-20220510035547-0e2f26c0a46a/pkg/pdpb/pdpb.pb.go:7691) +0x7b\n"
2022-07-18 09:27:03	log="[github.com/pingcap/kvproto/pkg/pdpb._PD_ScatterRegion_Handler.func1](http://github.com/pingcap/kvproto/pkg/pdpb._PD_ScatterRegion_Handler.func1)({0x33e4e10, 0xc003effec0}, {0x26bbf80?, 0xc0029449a0})\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/grpc_service.go:1246](http://github.com/pingcap/pd/server/grpc_service.go:1246) +0x193\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server.(*GrpcServer).ScatterRegion](http://github.com/tikv/pd/server.(*GrpcServer).ScatterRegion)(0xc000010538, {0x33e4e10?, 0xc003effec0?}, 0xc0029449a0)\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/grpc_service.go:1615](http://github.com/pingcap/pd/server/grpc_service.go:1615) +0x54\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server.scatterRegions](http://github.com/tikv/pd/server.scatterRegions)(0xc0012c2c80, {0xc0054d4660?, 0xc003effec0?, 0xc007e72060?}, {0xc0054d465e?, 0x3?}, 0xcd71ad?)\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/region_scatterer.go:200](http://github.com/pingcap/pd/server/schedule/region_scatterer.go:200) +0x5df\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule.(*RegionScatterer).ScatterRegionsByID](http://github.com/tikv/pd/server/schedule.(*RegionScatterer).ScatterRegionsByID)(0xc002ef5b60, {0xc0054d4660, 0x1, 0x1?}, {0xc0054d465e, 0x2}, 0x0?)\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/region_scatterer.go:224](http://github.com/pingcap/pd/server/schedule/region_scatterer.go:224) +0x30a\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule.(*RegionScatterer).scatterRegions](http://github.com/tikv/pd/server/schedule.(*RegionScatterer).scatterRegions)(0x24b2720?, 0xc003a633c0, 0x4b20?, {0xc0054d465e, 0x2}, 0x203000?)\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/region_scatterer.go:272](http://github.com/pingcap/pd/server/schedule/region_scatterer.go:272) +0x268\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule.(*RegionScatterer).Scatter](http://github.com/tikv/pd/server/schedule.(*RegionScatterer).Scatter)(0xc002ef5b60, 0xc0096cdc00, {0xc0054d465e, 0x2})\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/region_scatterer.go:319](http://github.com/pingcap/pd/server/schedule/region_scatterer.go:319) +0x5e3\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule.(*RegionScatterer).scatterRegion](http://github.com/tikv/pd/server/schedule.(*RegionScatterer).scatterRegion)(0xc002ef5b60, 0xc0096cdc00, {0xc0054d465e, 0x2})\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/region_scatterer.go:305](http://github.com/pingcap/pd/server/schedule/region_scatterer.go:305) +0x199\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule.(*RegionScatterer).scatterRegion.func1](http://github.com/tikv/pd/server/schedule.(*RegionScatterer).scatterRegion.func1)(0x24b2480?, {{0xc002ef6e80, 0x2, 0x2}, 0xc002ef6ea0, 0xc002ef6ec0})\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/region_scatterer.go:402](http://github.com/pingcap/pd/server/schedule/region_scatterer.go:402) +0x725\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule.(*RegionScatterer).selectCandidates](http://github.com/tikv/pd/server/schedule.(*RegionScatterer).selectCandidates)(0xc002ef5b60, 0xc007e72150?, 0x5, 0xc007e72150, {{0xc002ef6e80, 0x2, 0x2}, 0xc002ef6ea0, 0xc002ef6ec0})\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/filter/filters.go:126](http://github.com/pingcap/pd/server/schedule/filter/filters.go:126) +0x195\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule/filter.Target](http://github.com/tikv/pd/server/schedule/filter.Target)(0xc002ef6ea0?, 0xc002b89680, {0xc007d50360, 0x4, 0xc0027557c0?})\n"
2022-07-18 09:27:03	log="\t/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/pd/server/schedule/filter/filters.go:302](http://github.com/pingcap/pd/server/schedule/filter/filters.go:302) +0x59\n"
2022-07-18 09:27:03	log="[github.com/tikv/pd/server/schedule/filter.(*StoreStateFilter).Type(0xc00333b710?)\n](http://github.com/tikv/pd/server/schedule/filter.(*StoreStateFilter).Type(0xc00333b710?)\n)"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/print.go:219 +0x59\n"
2022-07-18 09:27:03	log="fmt.Sprintf({0x274b210, 0x15}, {0xc003a62718, 0x1, 0x1})\n"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/print.go:1026 +0x288\n"
2022-07-18 09:27:03	log="fmt.(*pp).doPrintf(0xc004179040, {0x274b210, 0x15}, {0xc003a62718?, 0x1, 0x1})\n"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/print.go:694 +0x60c\n"
2022-07-18 09:27:03	log="fmt.(*pp).printArg(0xc004179040, {0x23f6fa0?, 0xc00751fcb0}, 0x73)\n"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/print.go:446 +0xc5\n"
2022-07-18 09:27:03	log="fmt.(*pp).fmtString(0xc0012b8000?, {0x0?, 0x0?}, 0x7?)\n"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/format.go:359 +0x3f\n"
2022-07-18 09:27:03	log="fmt.(*fmt).fmtS(0xcda305?, {0x0?, 0xc005b3f304?})\n"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/format.go:110 +0x247\n"
2022-07-18 09:27:03	log="fmt.(*fmt).padString(0x1a7c2e0?, {0x0, 0xc})\n"
2022-07-18 09:27:03	log="\t/usr/local/go/src/fmt/print.go:82\n"
2022-07-18 09:27:03	log="fmt.(*buffer).writeString(...)\n"
2022-07-18 09:27:03	log="goroutine 243149 [running]:\n"
2022-07-18 09:27:03	log="\n"
2022-07-18 09:27:03	log="[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd39e90]\n"
2022-07-18 09:27:03	log="panic: runtime error: invalid memory address or nil pointer dereference\n"
```
### What did you see instead?
pd should normal

### What version of PD are you using (`pd-server -V`)?
./pd-server -V
 Release Version: v6.2.0-alpha
Edition: Community
Git Commit Hash: 51ad29c29ec23288cd66ad046e94d8975eaf4343
Git Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time:  2022-07-15 11:01:37

### PR DESCRIPTION
**Pull request by CabinfeverB**
Signed-off-by: Cabinfever_B <cabinfeveroier@gmail.com>

### What problem does this PR solve?

Issue Number: Close #5317

### What is changed and how does it work?
create new filter every scatter action

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/schedule/region_scatterer.go [hunk 1]
```diff
@@ -127,24 +127,30 @@ type RegionScatterer struct {
...
+		}),
 		specialEngines: make(map[string]engineContext),
 	}
 }
 
+type filterFunc func() filter.Filter
+
 type engineContext struct {
-	filters        []filter.Filter
+	filterFuncs    []filterFunc
```
Comment (rleungx) [inline server/schedule/region_scatterer.go:143]: Do we need to it in `engineContext`?

**Review by rleungx — APPROVED** [review]
The rest LGTM.

File: server/schedule/region_scatterer.go (hunk 1 above)
Comment (CabinfeverB) [reply server/schedule/region_scatterer.go:143]: This makes filters easy to use, I think we can keep it.

File: server/schedule/region_scatterer.go [hunk 2]
```diff
@@ -329,7 +335,9 @@ func (r *RegionScatterer) scatterRegion(region *core.RegionInfo, group string) *
 	for engine, peers := range specialPeers {
 		ctx, ok := r.specialEngines[engine]
 		if !ok {
-			ctx = newEngineContext(r.ctx, filter.NewEngineFilter(r.name, placement.LabelConstraint{Key: core.EngineKey, Op: placement.In, Values: []string{engine}}))
+			ctx = newEngineContext(r.ctx, func() filter.Filter {
+				return filter.NewEngineFilter(r.name, placement.LabelConstraint{Key: core.EngineKey, Op: placement.In, Values: []string{engine}})
+			})
```
Comment (AndreMouche) [inline server/schedule/region_scatterer.go:340]: could we use lock instead?

**Review by AndreMouche — COMMENTED** [review]
why do not use lock instead?

File: server/schedule/region_scatterer.go (hunk 2 above)
Comment (CabinfeverB) [reply server/schedule/region_scatterer.go:340]: IMO, the cost of lock is more than new

**Comment by AndreMouche** [comment]
Would it cause `GC` cost more with creating the filters and GC it frequently? How about comparing it with 
```
type RegionScatterer struct {
	……
	specialEngines sync.Map
}
type engineContext struct {
	filters        []filter.Filter
```

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**